### PR TITLE
Allow upload of signature for collection version.

### PR DESCRIPTION
**Pull request by rochacbruno**
# Description 🛠

Allow upload of signature for collection version.
    
- To upload a signature the endpoint is `/pulp/api/v3/content/ansible/collection_signatures/`
        The above endpoint expects:
         
        file: path/to/namespace.collection-version.asc (or any filename)
        signed_collection: href to CollectionVersion being signed
        repository:  href of the repository where that signature must be added

- To set a keyring to a repository

  `django-admin set-repo-keyring --repository=staging --keyring=/etc/pulp/certs/galaxy.kbx`

- To set the system to require a signature on approval

  `GALAXY_REQUIRE_SIGNATURE_FOR_APPROVAL=True`

then move endpoint will require at least one signature to allow approval.

Issue: AAH-1055

# Reviewer Checklists  👀
Developer reviewer:
- [ ] Code looks sound, good architectural decisions, no [code smells](https://www.codegrip.tech/productivity/everything-you-need-to-know-about-code-smells/)
- [ ] There is a Jira issue associated (note that "No-Issue" should be rarely used)
- [ ] Tests are included in `galaxy_ng/tests/integration` or `galaxy_ng/tests/functional`, and they fully cover necessary test scenarios… or tests not needed

QE reviewer ([exceptions](https://docs.engineering.redhat.com/display/AUTOHUB/Other+Team+Processes#OtherTeamProcesses-galaxy_ngrepo)):
- [ ] Tests are included in `galaxy_ng/tests/integration` or `galaxy_ng/tests/functional`, and they fully cover necessary test scenarios… or tests not needed
- [ ] PR meets applicable Acceptance Criteria for associated Jira issue

Note: when merging, include the Jira issue link in the squashed commit

### CODE REVIEW

**Comment by rochacbruno** [comment]
POST to upload signatures:

```bash
curl -X 'POST' \
  'http://0.0.0.0:5001/pulp/api/v3/content/ansible/collection_signatures/' \
  -H 'accept: application/json' \
  -H 'Authorization: Basic YWRtaW46YWRtaW4=' \
  -H 'Content-Type: multipart/form-data' \
  -H 'X-CSRFTOKEN: 4TqdnulLEhGgtp7QiHiFhk1rhlPtQQUbAwMJqc22H4i6hkIMJKrDy8LgAA23UA0s' \
  -F 'file=@MANIFEST.json.asc;type=text/plain' \
  -F 'repository=/pulp/api/v3/repositories/ansible/ansible/43dd716c-f7dd-4a80-b889-36818ed3e347/' \
  -F 'signed_collection=/pulp/api/v3/content/ansible/collection_versions/40f85531-29a5-4eb7-b4cd-a850f8824daf/'
``` 

```bash
curl -X 'POST' \
  'http://0.0.0.0:5001/pulp/api/v3/content/ansible/collection_signatures/' \
  -H 'accept: application/json' \
  -H 'Authorization: Basic YWRtaW46YWRtaW4=' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -H 'X-CSRFTOKEN: sH7uH2zhcCKJHC5fzszfCYhuyfV7v5ULYkt0KKgyfpmzvxGb0vIdTM1jRu8HzP02' \
  -d '{
  "file": {
    "name": "MANIFEST.json.asc",
    "type": "text/plain"
  },
  "repository": "/pulp/api/v3/repositories/ansible/ansible/43dd716c-f7dd-4a80-b889-36818ed3e347/",
  "signed_collection": "/pulp/api/v3/content/ansible/collection_versions/40f85531-29a5-4eb7-b4cd-a850f8824daf/"
}'

```

NOTES: 
### repository: takes the pulp_href for repository

which can be retrieved from:

```bash
curl -X 'GET' \
  'http://0.0.0.0:5001/pulp/api/v3/repositories/ansible/ansible/?name=staging' \
  -H 'accept: application/json' \
  -H 'Authorization: Basic YWRtaW46YWRtaW4='
```
```json
"results": [
    {
      "pulp_href": "/pulp/api/v3/repositories/ansible/ansible/9b772f40-48cb-43d9-9eaf-8201f9d8e3b5/",
      "pulp_created": "2022-03-31T17:05:59.247065Z",
      "versions_href": "/pulp/api/v3/repositories/ansible/ansible/9b772f40-48cb-43d9-9eaf-8201f9d8e3b5/versions/",
```

### signed_collection: takes pulp_href for signed collection

which is `/pulp/api/v3/content/ansible/collection_versions/<collection_version_id>`



After the upload is performed it can be verified on

GET http://0.0.0.0:5001/pulp/api/v3/content/ansible/collection_signatures/

```bash
 {
      "pulp_href": "/pulp/api/v3/content/ansible/collection_signatures/527ec05e-7873-44cc-9558-3f31a94a50d4/",
      "pulp_created": "2022-04-02T17:37:51.963061Z",
      "signed_collection": "/pulp/api/v3/content/ansible/collection_versions/40f85531-29a5-4eb7-b4cd-a850f8824daf/",
      "pubkey_fingerprint": "EBED170E8C9480E22A1D059B15250E9EC0A62577",
      "signing_service": null
    },
```

**Comment by rochacbruno** [comment]
@himdel @newswangerd tests on CI will not pass until we have API refactor and pulp_ansible 0.13 integrated.

To run locally you need: `.compose.env`
```bash
DEV_SOURCE_PATH='pulp_ansible:galaxy_ng'
LOCK_REQUIREMENTS=0
ENABLE_SIGNING=1
PULP_GALAXY_REQUIRE_CONTENT_APPROVAL=true
PULP_GALAXY_REQUIRE_SIGNATURE_FOR_APPROVAL=true
PULP_GALAXY_AUTO_SIGN_COLLECTIONS=false
```

Have your local clone of pulp ansible pointing to the latest commit that works with `galaxy_ng` master without api refactor.

> NOTE: Once API REfactor is merged, this is no more needed.

```bash
cd pulp_ansible/
git checkout 941bb15369c9f37a610e77b14644a1ec652bbe2e
```

Then, during the build of the containers the keyring must be set on the system.

One last step before you can start uploading signatures to `staging` is to set the field keyring=galaxy.kbx, as we don't have repo management yet this needs to be done via command line.

```bash
./compose exec api ./entrypoint.sh manage set-repo-keyring --repository staging --keyring galaxy.kbx          
This will set keyring to galaxy.kbx for staging repository, Proceed? (Y/n)Y
Process completed
Keyring galaxy.kbx set successfully to staging repository
```

Now you follow the endpoints on my previous [comment](https://github.com/ansible/galaxy_ng/pull/1197#issuecomment-1086691440) to upload the signatures.

**Comment by rochacbruno** [comment]
@himdel this zip contains a collection, a MANIFEST.json and a MANIFEST.json.asc (the signature to be uploaded) this was generated using the dev environment signing key, so must work for testing.

[signature_upload_assets.zip](https://github.com/ansible/galaxy_ng/files/8403357/signature_upload_assets.zip)

File: galaxy_ng/app/api/v3/viewsets/collection.py
Comment (newswangerd) [inline galaxy_ng/app/api/v3/viewsets/collection.py:485]: Needs i18n.

File: galaxy_ng/app/api/ui/views/public_key.py
Comment (newswangerd) [inline galaxy_ng/app/api/ui/views/public_key.py:82]: Repo lookups should always happen via a distribution

File: galaxy_ng/app/api/ui/urls.py
Comment (newswangerd) [inline galaxy_ng/app/api/ui/urls.py:183]: This set of paths and their corresponding views don't fit any of our standards.

1. Setting `base` as a url kwarg when there are only two options is confusing and makes the API way harder to navigate. It should have a separate set of paths for `repository` and `signing_service`
2. `PublicKeyView` is a viewset that's masquerading as a view. This should be rewritten as a viewset that has a `list` function for listing out the public keys and a `retrieve` function for retrieving a specific key. `api.base.ViewSet` should work for this since you've got a really weird set of querysets for this view and the base `ViewSet` doesn't provide any magic around pagination and viewsets
3. `PublicKeyView` doesn't declare any serializers or schemas, so it will break the API spec.

For the router you can do something like this. You could also make the file extension explicit as well, like you already have. 

```
[
   path(
        "",
        views.PublicKeyViewset.as_view({"get":"list"}),
        name='public-key',
    ),
    path(
        "repository/<path:file>",
        views.PublicKeyViewset.as_view({"get":"retrieve_repository"}),
        name='public-key-detail-key',
    ),
    path(
        "signing_service/<path:file>",
        views.PublicKeyViewset.as_view({"get":"retrieve_signing_service"}),
        name='public-key-detail-fingerprint',
    )
]
```

**Review by newswangerd — CHANGES_REQUESTED** [review]

File: galaxy_ng/app/api/ui/urls.py
Comment (rochacbruno) [reply galaxy_ng/app/api/ui/urls.py:183]: @newswangerd actually I will be removing this endpoint, just learned that we don't need to expose the public_key fingerprint for uploads anymore because the upload viewset can fetch it from signature.

Also, signing service public_key is already available on pulp/signing_Service/ API

File: galaxy_ng/app/api/v3/viewsets/collection.py
Comment (rochacbruno) [reply galaxy_ng/app/api/v3/viewsets/collection.py:485]: isn't the `_(`  already the i18n ?

Comment (newswangerd) [reply galaxy_ng/app/api/v3/viewsets/collection.py:485]: oops, I missed it. Sorry

**Review by newswangerd — APPROVED** [review]

**Review by awcrosby — APPROVED** [review]

File: docker/entrypoint.sh
Comment (rochacbruno) [reply docker/entrypoint.sh:101]: @awcrosby yes, keyring is for signature upload verification

**Comment by awcrosby** [comment]
@rochacbruno the `ci.int.devshift.net` pr_check.sh fails since automation-hub-galaxy-api pod does not get to a ready state. I spun up this PR image into an ephemeral env and see the pod error:

```
  File "/app/galaxy_ng/app/management/commands/maintain-pe-group.py", line 21, in handle
    pe_group, created = Group.objects.get_or_create(name=PE_GROUP_NAME)
...
django.db.utils.ProgrammingError: relation "auth_group" does not exist
LINE 1: ...ELECT "auth_group"."id", "auth_group"."name" FROM "auth_grou...
```

This _could_ be caused by the reverts related to RBAC, but I tried spinning an ephemeral env from master and it did not have any issues, could you try rebasing this PR and see if it fixes the error?

**Comment by rochacbruno** [comment]
@awcrosby rebased, but looks it is still failing on pr_check

**Comment by awcrosby** [comment]
I am going to disable this management command via https://github.com/ansible/galaxy_ng/pull/1214, and reenable it with RBAC changes, that should unblock this pr.

**Comment by awcrosby** [comment]
@rochacbruno I spun up an ephemeral env with these parameters 
```
GIT_COMMIT_INPUT=891d38be29ac32832483b7f4590eb3c655f88fc4
IMAGE_TAG_INPUT=pr-1197-a748cdd
```
I am seeing this before a pod shows a `CrashLoopBackOff` error and restarts and tries again, it never starts up and that is what is causing the `ci.int.devshift.net` timeouts:
```
/entrypoint.sh: ignoring /entrypoints.d/*
Setting up signing keyring.
gpg: directory '/tmp/ansible/.gnupg' created
gpg: keybox '/tmp/ansible/.gnupg/pubring.kbx' created
gpg: /tmp/ansible/.gnupg/trustdb.gpg: trustdb created
```

**Comment by rochacbruno** [comment]
@awcrosby @drodowic it is passing now, I added #1174 changes to clowder.yaml

**Review by brumik — APPROVED** [review]
I was able to get the UI working with these changes, so from usability perspective this looks good to me.

**Comment by rochacbruno** [comment]
@ironfroggy @newswangerd @awcrosby are we waiting something more to be able to merge this PR?

**Comment by newswangerd** [comment]
@rochacbruno I don't know. I'm happy to merge this now, but I don't know if this will cause problems for cloud.

**Comment by awcrosby** [comment]
Let’s wait until after the 26May cloud deploy is successful before merging this. On merge this will auto-deploy to stage, so we can have PE users interact with it there.

**Pull request merged**
